File: typescript/src/authz-service/adapters/graph/permissionModel.ts
// Permission model — pure data tables.
//
// Defines which relations are implied by others on the SAME object type.
// The graph evaluator reads these during traversal.
//
// Reading the tables: key = relation being checked, value = stronger relations
// that satisfy it.  WORKSPACE_RULES.viewer = ["editor"] means "workspace viewer
// is satisfied by workspace editor".

import type { Relation } from "../../../shared/rebac.ts";

export type ImpliedBy = Partial<Record<Relation, readonly Relation[]>>;

// team.admin implies team.member
export const TEAM_RULES: ImpliedBy = {
    member: ["admin"],
};

// workspace.owner implies workspace.editor implies workspace.viewer
export const WORKSPACE_RULES: ImpliedBy = {
    editor: ["owner"],
    viewer: ["editor"],
};

// document role hierarchy + computed permissions.
// owner/editor/viewer can ALSO be inherited from the parent workspace —
// that logic lives in makeGraphEvaluator.ts (requires a tuple lookup).
export const DOCUMENT_RULES: ImpliedBy = {
    can_read:    ["viewer"],
    can_comment: ["viewer"],
    can_edit:    ["editor"],
    can_delete:  ["owner"],
    viewer:      ["editor"],
    editor:      ["owner"],
};

File: go/internal/authzservice/adapters/graph/permissionmodel.go
package graph

import "rebac-primer/internal/shared"

// permissionmodel.go defines which relations are implied by others on the SAME
// object.  It mirrors typescript/src/authz-service/adapters/graph/permissionModel.ts.
//
// Reading the tables: key = relation being checked, value = stronger relations
// that satisfy it.
//
//	workspaceRules[viewer] = [editor]
//	→ "workspace viewer is satisfied by workspace editor"
//
// These are package-level vars allocated once, not on every call.

// impliedBy maps a relation to the set of stronger relations that imply it.
type impliedBy map[shared.Relation][]shared.Relation

// teamRules: team.admin implies team.member
var teamRules = impliedBy{
	shared.RelationTeamMember: {shared.RelationTeamAdmin},
}

// workspaceRules: workspace.owner ⊇ editor ⊇ viewer
var workspaceRules = impliedBy{
	shared.RelationWorkspaceEditor: {shared.RelationWorkspaceOwner},
	shared.RelationWorkspaceViewer: {shared.RelationWorkspaceEditor},
}

// documentRules: document role hierarchy and computed permissions.
// Note: owner/editor/viewer can also be inherited from the parent workspace —
// that logic lives in evaluator.go because it requires a tuple lookup.
var documentRules = impliedBy{
	shared.RelationDocumentCanRead:    {shared.RelationDocumentViewer},
	shared.RelationDocumentCanComment: {shared.RelationDocumentViewer},
	shared.RelationDocumentCanEdit:    {shared.RelationDocumentEditor},
	shared.RelationDocumentCanDelete:  {shared.RelationDocumentOwner},
	shared.RelationDocumentViewer:     {shared.RelationDocumentEditor},
	shared.RelationDocumentEditor:     {shared.RelationDocumentOwner},
}
